src/dataflow/src/logging/timely.rs

```diff
@@ -289,39 +286,65 @@ pub fn construct<A: Allocate>(
                 },
             );
 
+        let operates = operates.as_collection();
+
         // Accumulate the durations of each operator.
-        let elapsed = duration
+        let mut elapsed = duration
             .map(|(op, t, d)| (op, t, d as isize))
             .as_collection()
-            .count_total()
-            .map({
-                let mut row_packer = repr::RowPacker::new();
-                move |((id, worker), cnt)| {
-                    row_packer.pack(&[
-                        Datum::Int64(id as i64),
-                        Datum::Int64(worker as i64),
-                        Datum::Int64(cnt as i64),
-                    ])
-                }
-            });
+            .count_total();
 
-        let histogram = duration
+        // Accumulate histograms of execution times for each operator.
+        let mut histogram = duration
             .map(|(op, t, d)| ((op, d.next_power_of_two()), t, 1i64))
             .as_collection()
             .count_total()
-            .map({
-                let mut row_packer = repr::RowPacker::new();
-                move |(((id, worker), pow), cnt)| {
-                    row_packer.pack(&[
-                        Datum::Int64(id as i64),
-                        Datum::Int64(worker as i64),
-                        Datum::Int64(pow as i64),
-                        Datum::Int64(cnt as i64),
-                    ])
-                }
-            });
+            .map(|((key, pow), count)| ((key), (pow, count)));
+
+        let delay = std::time::Duration::from_nanos(10_000_000_000);
+
+        // Only keep metrics and histograms for the dataflow operators that currently exist.
+        elapsed = thin_collection(elapsed, delay, |c| {
+            operates.join_map(&c, move |key, _val1, val2| (*key, *val2))
```

Comment:
I think instead of join_map I want semijoin but not sure

Reply:
Yes, probably `c.semijoin(&operates)`. The main benefit would be a slightly leaner implementation (it knows not to use a complicated arrangement for `_val1` values).

Reply:
I ended up having to project away the value in operates to make this work but I think that since I had to do a map immediately after the join anyway it comes out to the same thing + a leaner arrangement